Sources/CoreLocationMiddleware/CoreLocationMiddleware.swift

```diff
@@ -67,43 +79,76 @@ public final class CoreLocationMiddleware: Middleware {
     }
+    
     public func handle(action: LocationAction, from dispatcher: ActionSource, afterReducer: inout AfterReducer) {
         switch action {
-        case .startMonitoring, .authorized:
-            startMonitoring()
-        case .stopMonitoring:
-            stopMonitoring()
+        case .startMonitoring: startMonitoring()
+        case .stopMonitoring: stopMonitoring()
+        case .requestAuthorizationStatus:
+            if #available(iOS 14.0, *) {
+                delegate.output?.dispatch(getAuthzStatus(status: manager.authorizationStatus))
+            } else {
+                return
+            }
+        case .requestAuthorizationType:
+            switch getState?().authzType {
+            case .always:
+                manager.requestAlwaysAuthorization()
+            case .whenInUse:
+                manager.requestWhenInUseAuthorization()
+            case .none:
+                break
+            }
+        case .requestPosition: manager.requestLocation()
+        case .requestDeviceCapabilities: delegate.output?.dispatch(getDeviceCapabilities())
         default: return
         }
     }
+    
     func startMonitoring() {
-        switch getState?() {
-        case .authorized?:
-            manager.startUpdatingLocation()
-        default:
-            // requestAlwaysAuthorization or requestWhenInUseAuthorization could be decided on
-            // the middleware init or as payload for startMonitoring
-            manager.requestAlwaysAuthorization()
+        let stateType = getState?().authzType
+        let stateStatus = getState?().authzStatus
+        
+        switch stateType {
+        case .always:
+            if CLAuthorizationStatus.authorizedAlways != stateStatus {
+                manager.requestAlwaysAuthorization()
+            }
+        case .whenInUse:
+            if ![CLAuthorizationStatus.authorizedAlways, CLAuthorizationStatus.authorizedWhenInUse].contains(stateStatus) {
+                manager.requestWhenInUseAuthorization()
+            }
+        case .none:
+            break
         }
+        manager.startUpdatingLocation()
     }
+    
     func stopMonitoring() {
         manager.stopUpdatingLocation()
     }
+    
+    func getDeviceCapabilities() -> LocationAction {
+        .gotDeviceCapabilities(
+            DeviceCapabilities(
+                isSignificantLocationChangeAvailable: CLLocationManager.significantLocationChangeMonitoringAvailable(),
+                isHeadingAvailable: CLLocationManager.headingAvailable(),
+                isRegionMonitoringAvailable: CLLocationManager.isMonitoringAvailable(for: CLCircularRegion.self),
+                isRangingAvailable: CLLocationManager.isRangingAvailable(),
+                isLocationServiceAvailable: CLLocationManager.locationServicesEnabled()
+            )
+        )
+    }
 }
 
- class CLDelegate: NSObject, CLLocationManagerDelegate {
+class CLDelegate: NSObject, CLLocationManagerDelegate {
```

Comment:
Very wise to separate this from the Middleware, so you can reuse this part from reactive chains and when EffectMiddleware is finally done. I liked!

Reply:
Thanks 🙏.
I initially had move the delegate implementation out of the CoreLocationMiddleware extension (44f31695eb2a8b710122dfdfdfe495dfa015fc29) but I don't remember exactly why (😬). However, yes, I do like that and I'll take the credit for it 🤣.